app/models/order.rb

```diff
@@ -225,6 +225,10 @@ def set_initial_state
     end
 
     before_transition on: :cancel do |order|
+      if OrdersPackage.where(order: order).any? { |op| op.dispatched? || op.dispatched_quantity.positive? }
+        raise Goodcity::InvalidStateError.new(I18n.t('orders_package.cancel_requires_undispatch'))
+      end
```

Comment:
Potential option 

`order.orders_package.where("state = 'dispatched' OR dispatched_quantity > 0").any?`

Reply:
Agree with @steveyken and more on same: 

order.orders_package.where("state = 'dispatched' OR dispatched_quantity > 0").exists? will be better in some cases where we have large number of data.

Here we might not need exists but thought to share it.

https://www.ombulabs.com/blog/benchmark/performance/rails/present-vs-any-vs-exists.html

Reply:
Will update it, although that's just one node of a greater web of potential issues.

Just a few lines below you'll notice `order.orders_packages.each` which has the same potential issues.

If you suspect performance issues I recommend figuring out a general approach to handling them (we can create a ticket for that). Singular changes here and there won't matter enough to make an impact